[FEATURE] Direct web app (ASGI/WSGI) testing from python tests

It seems that python tests using `schemathesis.from_asgi` are run using Requests, and expect a server to be running. I would like to run the tests directly as happens with `schemathesis run --app` on the command line. If this is already possible I was not able to figure out how by reading the docs and searching the code.

Thanks for your reply, @Stranger6667!

My setup is very similar to the one in the #1305 bug you linked. I am getting the following error:

```
>           raise MissingSchema(error)
E           requests.exceptions.MissingSchema: Invalid URL '/health': No schema supplied. Perhaps you meant http:///health?

../__pypackages__/3.9/lib/requests/models.py:390: MissingSchema
```

This error from requests was deeply confusing at first because the library has a typo and means "scheme" (eg `http`) when it says "schema." This doesn't appear to be a schema error. All my endpoints produce this same error.

If this isn't enough info, I will try to provide a minimized example for you. Thanks again for your help and for this useful library.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -4,6 +4,10 @@ Changelog
 `Unreleased`_ - TBD
 -------------------
 
+**Changed**
+
+- Validate ``requests`` kwargs to catch cases when the ASGI integration is used, but the proper ASGI client is not supplied. `#1335`_
+
 `3.13.3`_ - 2022-02-20
 ----------------------
 
@@ -2339,6 +2343,7 @@ Deprecated
 .. _#1342: https://github.com/schemathesis/schemathesis/issues/1342
 .. _#1340: https://github.com/schemathesis/schemathesis/issues/1340
 .. _#1336: https://github.com/schemathesis/schemathesis/issues/1336
+.. _#1335: https://github.com/schemathesis/schemathesis/issues/1335
 .. _#1331: https://github.com/schemathesis/schemathesis/issues/1331
 .. _#1328: https://github.com/schemathesis/schemathesis/issues/1328
 .. _#1325: https://github.com/schemathesis/schemathesis/issues/1325

--- a/docs/python.rst
+++ b/docs/python.rst
@@ -401,6 +401,24 @@ in this case, the test execution will go much faster.
         response = case.call_wsgi()
         case.validate_response(response)
 
+If you don't supply the ``app`` argument to the loader, make sure you pass your test client when running tests:
+
+.. code-block:: python
+
+    @pytest.fixture()
+    def app_schema(client):
+        openapi = client.app.openapi()
+        return schemathesis.from_dict(openapi)
+
+
+    schema = schemathesis.from_pytest_fixture("app_schema")
+
+
+    @schema.parametrize()
+    def test_api(case, client):
+        # The `session` argument must be supplied.
+        case.call_and_validate(session=client)
+
 Unittest support
 ----------------
 

--- a/src/schemathesis/models.py
+++ b/src/schemathesis/models.py
@@ -27,7 +27,7 @@ from typing import (
     Union,
     cast,
 )
-from urllib.parse import quote, unquote, urljoin, urlsplit, urlunsplit
+from urllib.parse import quote, unquote, urljoin, urlparse, urlsplit, urlunsplit
 
 import attr
 import curlify
@@ -314,14 +314,15 @@ class Case:  # pylint: disable=too-many-public-methods
         **kwargs: Any,
     ) -> requests.Response:
         """Make a network call with `requests`."""
+        data = self.as_requests_kwargs(base_url, headers)
+        data.update(kwargs)
+        data.setdefault("timeout", DEFAULT_RESPONSE_TIMEOUT / 1000)
         if session is None:
+            validate_vanilla_requests_kwargs(data)
             session = requests.Session()
             close_session = True
         else:
             close_session = False
-        data = self.as_requests_kwargs(base_url, headers)
-        data.update(kwargs)
-        data.setdefault("timeout", DEFAULT_RESPONSE_TIMEOUT / 1000)
         try:
             response = session.request(**data)  # type: ignore
         except requests.Timeout as exc:
@@ -471,6 +472,21 @@ class Case:  # pylint: disable=too-many-public-methods
             cookies=deepcopy(self.cookies),
             query=deepcopy(self.query),
             body=deepcopy(self.body),
+        )
+
+
+def validate_vanilla_requests_kwargs(data: Dict[str, Any]) -> None:
+    """Check arguments for `requests.Session.request`.
+
+    Some arguments can be valid for cases like ASGI integration, but at the same time they won't work for the regular
+    `requests` calls. In such cases we need to avoid an obscure error message, that comes from `requests`.
+    """
+    url = data["url"]
+    if not urlparse(url).netloc:
+        raise RuntimeError(
+            "The URL should be absolute, so Schemathesis knows where to send the data. \n"
+            f"If you use the ASGI integration, please supply your test client "
+            f"as the `session` argument to `call`.\nURL: {url}"
         )
 
 

--- a/test/test_models.py
+++ b/test/test_models.py
@@ -204,6 +204,19 @@ def test_call_and_validate(openapi3_schema_url):
     test()
 
 
+@pytest.mark.operations("success")
+def test_call_asgi_and_validate(fastapi_app):
+    api_schema = schemathesis.from_dict(fastapi_app.openapi())
+
+    @given(case=api_schema["/users"]["GET"].as_strategy())
+    @settings(max_examples=1)
+    def test(case):
+        with pytest.raises(RuntimeError, match="The URL should be absolute"):
+            case.call_and_validate()
+
+    test()
+
+
 def test_case_partial_deepcopy(swagger_20):
     operation = APIOperation("/example/path", "GET", {}, swagger_20)
     original_case = Case(

